File: BombSquad_MapGen/Region.js
/*
 *      Name: Region
 *      Author: Mitchell Croft
 *      Created: 04/11/2017
 *      Modified: 05/11/2017
 * 
 *      Purpose:
 *      Monitor a collection of coordinates as a Region on a 2D string map
 *      and at request find the distance of each coord from an edge
 **/

/*
    Region : Constructor - Initialise with a collection of values
    Created: 04/11/2017
    Modified: 04/11/2017

    param[in] pCoords - An array of Vec2 objects containing enclosed
    param[in] pWidth - The width of the map the coordinates are on
    param[in] pHeight - The height of the map the coordinates are on
*/
function Region(pCoords, pWidth, pHeight) {
    //! Store the dimenions of the map
    this.mapWidth = pWidth;
    this.mapHeight = pHeight;

    //! Store the coordinates in the region
    this.coordinates = [];

    //! Store the 'center of mass' for the region
    this.center = new Vec2();

    //! Store the regions that are linked together
    this.links = [];

    //! Store a hitmap of the different coordinates
    this.hitMap = [];

    //! Store a map of the distances of each coord from an edge
    this.depthMap = [];

    //! Loop through and read from the parameter
    for (let i = 0; i < pCoords.length; i++) {
        //! Get the coordinate value to use
        let coord = pCoords[i];

        //! Store the cooridinate in the internal map
        this.coordinates[i] = coord;

        //! Add to the center of mass
        this.center.addSet(coord);

        //! Check the index exists in the info maps
        if (!(this.hitMap[coord.x] instanceof Array)) {
            this.hitMap[coord.x] = [];
            this.depthMap[coord.x] = [];
        }

        //! Register the point in the hitMap
        this.hitMap[coord.x][coord.y] = "X";

        //! Assign the default depth value
        this.depthMap[coord.x][coord.y] = 0;
    }

    //! Average out the center of mass
    this.center.divSet(this.coordinates.length);

    //! Store a list of coordinates on the edge of the region
    this.edges = [];

    //! Store a list of the axis offsets to check for edges
    let edgeCheckOffsets = [new Vec2(-1, 0), new Vec2(1, 0), new Vec2(0, -1), new Vec2(0, 1)];

    //! Loop through the coordinates and check if its on the edge
    for (let i = 0; i < this.coordinates.length; i++) {
        //! Get the coordinate
        let coord = this.coordinates[i];

        //! Check if coordinate is on the bounds of the map
        if (coord.x === 0 || coord.x === pWidth - 1 ||
            coord.y === 0 || coord.y === pHeight - 1);

        //! Check if the coordinate has (at least) one non-contained coordinate
        else {
            //! Loop through the axis offset values
            let found = false;
            for (let j = 0; j < edgeCheckOffsets.length; j++) {
                //! The offset position to check the hitmap at
                let offset = coord.add(edgeCheckOffsets[j]);

                //! Check the offset coords are on the map
                if (offset.x < 0 || offset.y < 0 || offset.x >= pWidth || offset.y >= pHeight)
                    continue;

                //! Ensure the check bounds on the 'x' axis are valid arrays
                else if (!(this.hitMap[offset.x] instanceof Array))
                    this.hitMap[offset.x] = [];

                //! Check if the offset coordinate is not on the hit map
                if (this.hitMap[offset.x][offset.y] !== "X") {
                    found = true;
                    break;
                }
            }

            //! Check if an edge was found
            if (!found) continue;
        }

        //! Add the coordinate to the edges list
        this.edges.push(coord);
    }
};

/*
    Region : analyseDepth - Flood search each coordinate to find the distance from an edge coordinate
    Created: 04/11/2017
    Modified: 05/11/2017
*/
Region.prototype.analyseDepth = function() {
    //Create a temporary coordinate element that tracks position data and the object which added it to the search list
    function CoordTrace(pVec, pParent) {
        //Store the coordinates
        this.pos = Validate.instance(pVec, Vec2, null, true);

        //Store the parent of the coord
        this.parent = Validate.instance(pParent, CoordTrace, null);
    };

    //Store the offsets used to find neighbours
    let offsets = [new Vec2(-1, 0), new Vec2(1, 0), new Vec2(0, -1), new Vec2(0, 1), new Vec2(-1, -1), new Vec2(-1, 1), new Vec2(1, 1), new Vec2(1, -1)];

    //Loop through the coordinates to find their depth
    for (let i = 0; i < this.coordinates.length; i++) {
        //Store the coordinate to retrieve the value of
        let coord = this.coordinates[i];

        //Store a list of the Coord Traces to investigate
        let unsearched = [];

        //Store a map of the searched coordinates
        let searchMap = [];

        //Add the current coordinate to the list
        unsearched.push(new CoordTrace(coord));

        //Register the first coordiante
        searchMap[coord.x] = [];
        searchMap[coord.x][coord.y] = 1;

        //Flag if the search has gone out of bounds
        let outOfBounds = false;

        //Loop through and process trace coords
        let current = null;
        while (current = unsearched.shift()) {
            //Check to see if this point is not on the hitmap
            if (this.hitMap[current.pos.x][current.pos.y] !== "X")
                break;

            //Add neighbouring cells to the unsearched list
            for (let j = 0; j < offsets.length; j++) {
                //Get the offset position
                let offset = current.pos.add(offsets[j]);

                //Check the coordinate is in bounds
                if (offset.x < 0 || offset.y < 0 ||
                    offset.x >= this.mapWidth || offset.y >= this.mapHeight) {
                    outOfBounds = true;
                    break;
                }

                //Check the position is valid on the searched map
                if (!(searchMap[offset.x] instanceof Array))
                    searchMap[offset.x] = [];

                //Check if offset coordinate has been searched
                if (searchMap[offset.x][offset.y] !== 1) {
                    //Add the coordinate to the search list
                    unsearched.push(new CoordTrace(offset, current));

                    //Flag coordinate on searched map
                    searchMap[offset.x][offset.y] = 1;
                }
            }

            //If search has gone out fo bounds break
            if (outOfBounds) break;
        }

        //Check to see if the edge was found
        if (current) {
            //Reset the depth value
            this.depthMap[coord.x][coord.y] = -1;

            //Trace back path
            while (current !== null) {
                this.depthMap[coord.x][coord.y]++;
                current = current.parent;
            }
        }
    }
};

File: SDL2_Engine/Projects/BombSquad/src/Game/Map Generation/Region.cpp
#include "Region.hpp"

//! Include the required STL objects
#include <queue>

namespace BombSquad {
	/*
		Region : Constructor - Initialise with default values
		Created: 08/11/2017
		Modified: 08/11/2017

		param[in] pCoords - A collection of coords that are contained within the region
		param[in] pWidth - The width of the map that the coordinates exist on
		param[in] pHeight - The height of the map that the coordinates exist on
	*/
	Region::Region(const coords& pCoords, const size_t& pWidth, const size_t& pHeight) : contained(pCoords) {
		//Create the location map
		locationMap.setBounds(pWidth, pHeight);

		//Loop through and process the contained coordinates
		for (size_t i = 0; i < contained.size(); i++) {
			//Add the coordinate to the center of mass
			center += contained[i];

			//Register the coordinate in the location map
			locationMap[contained[i].x][contained[i].y] = 1;
		}

		//Take the average to get the center of mass
		center /= contained.size();

		//Mark the offsets to apply to search for outer-region areas
		const coord OFFSETS[] = { {-1, 0}, {1, 0}, {0, -1}, {0, 1} };
		const size_t OFFSET_SIZE = sizeof(OFFSETS) / sizeof(const coord);

		//Loop through the coordinates
		for (size_t i = 0; i < contained.size(); i++) {
			//Get a reference to the current coordinate value
			const coord& COORD = contained[i];

			//Check if the coord is on the bounds of the map
			if (COORD.x == 0 || COORD.y == 0 ||
				COORD.x == pWidth - 1 || COORD.y == pHeight - 1);

			//Check if the coordinate has (at least) one non-contained coordinate
			else {
				//Loop through the axis offset values
				bool found = false;
				for (size_t j = 0; j < OFFSET_SIZE; j++) {
					//Apply the offset position
					const coord OFFSET = COORD + OFFSETS[j];

					//Check the coord is on map
					if (OFFSET.x < 0 || OFFSET.y < 0 ||
						OFFSET.x >= pWidth || OFFSET.y >= pHeight)
						continue;

					//Check if the coordinate is on location map
					else if (!locationMap[OFFSET.x][OFFSET.y]) {
						found = true;
						break;
					}
				}

				//Check if an external coordinate was found
				if (!found) continue;
			}

			//Add the coordinate to the edges list
			edges.push_back(COORD);
		}
	}

	/*
		Region : analyseDepth - Flood search each coordinate to find the distance from a coordinate not contained in the region
		Created: 08/11/2017
		Modified: 08/11/2017
	*/
	void Region::analyseDepth() {
		//Create a temporary coordinate element that tracks position and parent object data
		struct CoordTrace {
			//! Store the position
			coord pos;

			//! Store the parent of the coordinate
			CoordTrace* parent = nullptr;
		};

		//Setup the depth map
		distanceFromExternal.setBounds(locationMap.width(), locationMap.height());

		//Store the offsets used to find neighbors
		const coord OFFSETS[] = { {-1, 0}, {1, 0}, {0, -1}, {0, 1}, {-1, -1}, {-1, 1}, {1, 1}, {1, -1} };
		const size_t OFFSET_SIZE = sizeof(OFFSETS) / sizeof(const coord);

		//Loop through all coordinates to find their depth
		for (size_t i = 0; i < contained.size(); i++) {
			//Store a reference to the current coordinate being processed
			const coord COORD = contained[i];

			//Create a hit map for searched coordinates
			hitMap searched(locationMap.width(), locationMap.height());

			//Store a queue of coordinates to process
			std::queue<CoordTrace*> unsearched;

			//Create a CoordTrace Grid
			Map<CoordTrace> traceGrid(locationMap.width(), locationMap.height());

			//Set the position of the initial coord
			traceGrid[COORD.x][COORD.y].pos = COORD;

			//Add the current coordinate to the queue
			unsearched.push(&traceGrid[COORD.x][COORD.y]);

			//Flag if the search has gone out of bounds
			bool outOfBounds = false;

			//Loop through and process trace coords
			CoordTrace* current = nullptr;
			while (unsearched.size()) {
				//Get the next item
				current = unsearched.front();
				unsearched.pop();

				//Check to see if this location is outside of the region
				if (!locationMap[current->pos.x][current->pos.y])
					break;

				//Add neighboring cells to the unsearched list
				for (size_t j = 0; j < OFFSET_SIZE; j++) {
					//Get the offset position
					const coord OFFSET = current->pos + OFFSETS[j];

					//Check the coordinate is in bounds
					if (OFFSET.x < 0 || OFFSET.y < 0 ||
						OFFSET.x >= searched.width() || OFFSET.y >= searched.height()) {
						outOfBounds = true;
						break;
					}

					//Check if the coordinate has been searched
					else if (!searched[OFFSET.x][OFFSET.y]) {
						//Setup the Grid Node
						traceGrid[OFFSET.x][OFFSET.y].pos = OFFSET;
						traceGrid[OFFSET.x][OFFSET.y].parent = current;

						//Add the offset into the unsearched queue
						unsearched.push(&traceGrid[OFFSET.x][OFFSET.y]);

						//Flag the offset in the searched map
						searched[OFFSET.x][OFFSET.y] = 1;
					}
				}

				//If search went out of bounds, break
				if (outOfBounds) break;
			}

			//Check to see if an outer-region was found
			if (current) {
				//Reset the depth value
				distanceFromExternal[COORD.x][COORD.y] = -1;

				//Trace the path back
				while (current) {
					//Increase the distance
					distanceFromExternal[COORD.x][COORD.y]++;

					//Move up the chain
					current = current->parent;
				}
			}
		}
	}

	/*
		Region : recursiveLinkToMain - Recursively traverse the links vector and toggle the linkedToMain
									   flag
		Created: 09/11/2017
		Modified: 09/11/2017
	*/
	void Region::recursiveLinkToMain() {
		//Check if the current is already linked to main
		if (linkedToMain) return;

		//Toggle the linked flag
		linkedToMain = true;

		//Recurse through the links
		for (size_t i = 0; i < links.size(); i++)
			links[i]->recursiveLinkToMain();
	}
}